akka-typed/src/main/scala/akka/typed/javadsl/ReceiveBuilder.scala

```diff
@@ -0,0 +1,170 @@
+   * @tparam M type of signal to match
+   * @return a new behavior with the specified handling appended
+   */
+  def onSignal[M <: Signal](`type`: Class[M], handler: Function[M, Behavior[T]]): ReceiveBuilder[T] =
+    withSignal(`type`, None, signal ⇒ handler.apply(signal.asInstanceOf[M]))
+
+  /**
+   * Add a new predicated case to the signal handling.
+   *
+   * @param type type of signals to match
+   * @param test a predicate that will be evaluated on the argument if the type matches
+   * @param handler action to apply if the type matches and the predicate returns true
+   * @tparam M type of signal to match
+   * @return a new behavior with the specified handling appended
+   */
+  def onSignal[M <: Signal](`type`: Class[M], test: Predicate[M], handler: Function[M, Behavior[T]]): ReceiveBuilder[T] =
+    withSignal(
+      `type`,
+      Some((t: Signal) ⇒ test.test(t.asInstanceOf[M])),
+      signal ⇒ handler.apply(signal.asInstanceOf[M])
+    )
+
+  /**
+   * Add a new case to the signal handling without compile time type check.
+   *
+   * Should normally not be used, but when matching on class with generic type
+   * argument it can be useful, e.g. <code>GenMsg.class</code> and <code>(ActorContext<Message> ctx, GenMsg&lt;String&gt; list) -> {...}</code>
+   *
+   * @param type type of signal to match
+   * @param handler action to apply when the type matches
+   * @return a new behavior with the specified handling appended
+   */
+  def onSignalUnchecked[M <: Signal](`type`: Class[_ <: Signal], handler: Function[M, Behavior[T]]): ReceiveBuilder[T] =
+    withSignal(`type`, None, signal ⇒ handler.apply(signal.asInstanceOf[M]))
+
+  /**
+   * Add a new case to the signal handling matching equal signals.
+   *
+   * @param signal the signal to compare to
+   * @param handler action to apply when the message matches
+   * @return a new behavior with the specified handling appended
+   */
+  def onSignalEquals(signal: Signal, handler: Creator[Behavior[T]]): ReceiveBuilder[T] =
+    withSignal(signal.getClass, Some(_.equals(signal)), _ ⇒ handler.create())
+
+  private def withMessage(`type`: Class[_ <: T], test: Option[T ⇒ Boolean], handler: T ⇒ Behavior[T]): ReceiveBuilder[T] =
+    new ReceiveBuilder[T](Case[T, T](`type`, test, handler) +: messageHandlers, signalHandlers)
+
+  private def withSignal[M <: Signal](`type`: Class[M], test: Option[Signal ⇒ Boolean], handler: Signal ⇒ Behavior[T]): ReceiveBuilder[T] =
+    new ReceiveBuilder[T](messageHandlers, Case[T, Signal](`type`, test, handler) +: signalHandlers)
+}
+
+object ReceiveBuilder {
+  def create[T]: ReceiveBuilder[T] = new ReceiveBuilder[T](Nil, Nil)
+
+  import scala.language.existentials
+
+  /** INTERNAL API */
+  @InternalApi
+  private[javadsl] final case class Case[BT, MT](`type`: Class[_ <: MT], test: Option[MT ⇒ Boolean], handler: MT ⇒ Behavior[BT])
+
+}
+
+/**
+ * Receive type for [[Actor.MutableBehavior]]
+ */
+class Receive[T] private[javadsl] (
```

Comment:
make it a pure interface right away (trait with only abstract methods `receiveMessage` and `receiveSignal`, because then we open up for other implementations such as `VAVr`.

Thinking about where this should be located. Would be nice with as few imports as possible. For `AbstractActor` we placed it in the `AbstractActor` companion for that reason. Try the same.